### PR TITLE
fix: Guard Minuit optimizer against provided strategy of None

### LINKED ISSUE #1785: Documentation of default Minuit strategy choice
### Summary

The documentation of `minuit_optimizer` lists `strategy` as a possible kwarg with default value of `None`, but it does not explain what that choice leads to. `pyhf` uses `0`/`1` depending on whether user-provided gradients are used (#1172 / #1183), and I believe it would be good to have that documented on the optimizer page. I'd be happy to submit a PR.

### Documentation Page Link

https://pyhf.readthedocs.io/en/v0.6.3/_generated/pyhf.optimize.opt_minuit.minuit_optimizer.html#pyhf.optimize.opt_minuit.minuit_optimizer

### Code of Conduct

- [X] I agree to follow the Code of Conduct

### PR DESCRIPTION
**Pull request by matthewfeickert**
# Description

* Fixes #1785
* Amends PR https://github.com/scikit-hep/pyhf/pull/2277

Guard Minuit optimizer strategy from None by first attempting to get a not-None result from the options kwargs and then falling back to assigning the strategy through `int(not do_grad)`. In the process this simplifies the logic required and supersedes PR https://github.com/scikit-hep/pyhf/pull/2277.

Add tests to `test_minuit_strategy_global` for `strategy=None` and `strategy=2` (`2` is a valid `iminuit.Minuit.strategy` strategy but not used often, so done just for good measure).

Add a very brief explanation to docs of what using `strategy=None` means in terms of evaluation to 0 or 1 from `int(pyhf.tensorlib.default_do_grad)`.

This extended example from @alexander-held in Issue #1785 now runs with this PR:

```python
# issue-1785.py
import pyhf

model = pyhf.simplemodels.uncorrelated_background(
    signal=[12.0, 11.0], bkg=[50.0, 52.0], bkg_uncertainty=[3.0, 7.0]
)
data = [51, 48] + model.config.auxdata
pyhf.set_backend("numpy", "minuit")
pyhf.infer.mle.fit(data, model)

strategies = [None, 0, 1, 2]
for strategy in strategies:
    pyhf.infer.mle.fit(data, model, strategy=strategy)
```

# Checklist Before Requesting Reviewer

- [x] Tests are passing
- [x] "WIP" removed from the title of the pull request
- [x] Selected an Assignee for the PR to be responsible for the log summary

# Before Merging

For the PR Assignees:

- [x] Summarize commit messages into a comprehensive review of the PR

```
* Guard Minuit optimizer strategy from None by first attempting to get a
  not-None result from the options kwargs and then falling back to
  assigning the strategy through the equivalent of `int(not do_grad)`.
  In the process this simplifies the logic required and supersedes PR
  https://github.com/scikit-hep/pyhf/pull/2277.
   - Cast the truthiness of `do_grad` as an integer for human
     readability in debugging.
* Update the comments on `do_grad` strategy with definitions from
  iMinuit v2.24.0 docs.
* Add tests to test_minuit_strategy_global for strategy=None and strategy=2
  (2 is a valid iminuit.Minuit.strategy strategy but not used often).
* Add brief explanation to docs of what using strategy=None means in
  terms of evaluation to 0 or 1 from `int(pyhf.tensorlib.default_do_grad)`.
```

### CODE REVIEW

File: src/pyhf/optimize/opt_minuit.py [hunk 1]
```diff
@@ -102,8 +105,11 @@ def _minimize(
         # 0: Fast, user-provided gradient
         # 1: Default, no user-provided gradient
         strategy = options.pop(
-            'strategy', self.strategy if self.strategy is not None else not do_grad
+            'strategy', self.strategy if self.strategy is not None else int(not do_grad)
```
Comment (matthewfeickert) [inline src/pyhf/optimize/opt_minuit.py:108]: As the `iminuit.Minuit.strategy` is an integer, instead of having to have a human determine the integer representation of the inverse of the truthiness of `do_grad` during debugging or inspection do this in code.

File: src/pyhf/optimize/opt_minuit.py [hunk 2]
```diff
@@ -102,8 +105,11 @@ def _minimize(
...
         )
+        # Passing strategy=None to options requires another check to guard against None
+        if strategy is None:
+            strategy = int(not do_grad)
```
Comment (matthewfeickert) [inline src/pyhf/optimize/opt_minuit.py:112]: This is done _after_ the above and not _before_ by using

```python
        if self.strategy is None:
            self.strategy = int(not do_grad)
```

to avoid mutating `self.strategy` making it harder to trace during debugging.

File: tests/test_optim.py [hunk 3]
```diff
@@ -173,27 +173,35 @@ def test_minuit_strategy_do_grad(mocker, backend):
     assert spy.spy_return.minuit.strategy == 1
 
 
-@pytest.mark.parametrize('strategy', [0, 1])
+@pytest.mark.parametrize('strategy', [0, 1, 2])
```
Comment (matthewfeickert) [inline tests/test_optim.py:176]: Adding `2` as we can and I forgot it existed.

**Comment by matthewfeickert** [comment]
Relevant RTD build: https://pyhf--2278.org.readthedocs.build/en/2278/_generated/pyhf.optimize.opt_minuit.minuit_optimizer.html

^ @alexander-held

File: src/pyhf/optimize/opt_minuit.py [hunk 4]
```diff
@@ -28,7 +28,10 @@ def __init__(self, *args, **kwargs):
         Args:
             errordef (:obj:`float`): See minuit docs. Default is ``1.0``.
             steps (:obj:`int`): Number of steps for the bounds. Default is ``1000``.
-            strategy (:obj:`int`): See :attr:`iminuit.Minuit.strategy`. Default is ``None``.
+            strategy (:obj:`int`): See :attr:`iminuit.Minuit.strategy`.
+              Default is ``None``, which results in either
+              :attr:`iminuit.Minuit.strategy` ``0`` or ``1`` from the evaluation of
+              ``int(not pyhf.tensorlib.default_do_grad)``.
```
Comment (matthewfeickert) [inline src/pyhf/optimize/opt_minuit.py:34]: c.f. https://pyhf--2278.org.readthedocs.build/en/2278/_generated/pyhf.optimize.opt_minuit.minuit_optimizer.html for render.

**Review by kratsg — APPROVED** [review]

**Comment by matthewfeickert** [comment]
Once this is approved I'll make patch release `pyhf` `v0.7.3` with this fix included in it.

File: src/pyhf/optimize/opt_minuit.py [hunk 5]
```diff
@@ -99,11 +102,14 @@ def _minimize(
...
-            'strategy', self.strategy if self.strategy is not None else not do_grad
-        )
+        # int(not do_grad) results in iminuit.Minuit.strategy of either:
+        #   0: Fast. Does not check a user-provided gradient.
+        #   1: Default. Checks user-provided gradient against numerical gradient.
+        strategy = options.pop("strategy", self.strategy)
+        # Guard against None from either self.strategy defaulting to None or
+        # passing strategy=None as options kwarg
+        if strategy is None:
+            strategy = int(not do_grad)
```
Comment (matthewfeickert) [inline src/pyhf/optimize/opt_minuit.py:112]: As the `iminuit.Minuit.strategy` is an integer, instead of having to have a human determine the integer representation of the inverse of the truthiness of `do_grad` during debugging or inspection do this in code.

Comment (alexander-held) [reply src/pyhf/optimize/opt_minuit.py:112]: If you're aiming for maximum readability, I like `strategy = 0 if do_grad else 1` best.

**Review by alexander-held — COMMENTED** [review]
This looks good to me, thanks! Just a tiny comment on what I believe is slightly more readable, but might be preference.

As a heads-up: there will also be a strategy=3 arriving eventually: https://github.com/root-project/root/pull/13109 (more expensive and stable).
